File: tests/svm-agreement-registry.ts
import * as anchor from "@coral-xyz/anchor";
import { BorshCoder, Idl, Program, web3 } from "@coral-xyz/anchor";
import { signBytes, createKeyPairFromBytes, generateKeyPair, getBase58Decoder } from "@solana/kit";
import { SvmAgreementRegistry } from "../target/types/svm_agreement_registry";
import { expect } from "chai"
import { ethers } from "ethers"

describe("svm-agreement-registry", () => {
  // Configure the client to use the local cluster.
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  const program = anchor.workspace.svmAgreementRegistry as Program<SvmAgreementRegistry>;

  it("accepts Ed25519 signatures", async () => {
    const kvPairs: {key: string, value: string}[] = [
      { key: "name", value: "Alice"},
      { key: "age", value: "30"},
    ];

    // Format and sign the message

    const serializedKvPairs = serializeKeyValuePairs(kvPairs);

    // Construct the message preamble as per Solana CLI (https://github.com/anza-xyz/solana-sdk/blob/fed203a9e8b2a9d3b1a0c98c004e3d1ede569d0c/offchain-message/src/lib.rs#L184-L190)
    // TODO uncomment the missing fields once solana-cli has finally followed the spec (PR: https://github.com/anza-xyz/agave/issues/3340)
    const signingDomain = Buffer.from("\xffsolana offchain", "ascii");
    const headerVersion = Buffer.from([0]); // Version 0
    // const applicationDomain = Buffer.alloc(32); // Arbitrary 32-byte domain, can customize
    const messageFormat = Buffer.from([0]); // Message format 0 (printable chars)
    // const numSigners = Buffer.from([1]); // Single signer
    // const signerPubkey = provider.wallet.publicKey.toBuffer(); // 32-byte public key
    // const messageBytes = Buffer.from(message, "ascii");
    const messageLength = Buffer.alloc(2);
    messageLength.writeUInt16LE(serializedKvPairs.length); // Little-endian u16

    // Combine all parts into the formatted offchain message
    const offchainMessage = Buffer.concat([
      signingDomain,
      headerVersion,
      // applicationDomain,
      messageFormat,
      // numSigners,
      // signerPubkey,
      messageLength,
      serializedKvPairs,
    ]);

    // Sign it
    const keyPair = await createKeyPairFromBytes(provider.wallet.payer.secretKey);
    const signatureBytes = await signBytes(keyPair.privateKey, new Uint8Array(offchainMessage));
    const signature = Array.from(signatureBytes);

    // TODO test: uncomment to verify signature locally
    // const verified = await verifySignature(keyPair.publicKey, signedBytes, finalMessage);
    // console.log({ verified });

    const newDataEntryAccountKp = new web3.Keypair();
    const tx = (new web3.Transaction())
      // SvmAgreementRegistry will check if the previous instruction is a valid Ed25519 instruction, so we must call it
      // immediately before calling SvmAgreementRegistry.
      .add(
        anchor.web3.Ed25519Program.createInstructionWithPublicKey({
          // TODO Is seems to work but the error returned is weird:
          //  Error processing Instruction 0: custom program error: 0x2
          // publicKey: newDataEntryAccountKp.publicKey.toBytes(),
          // message: messageBytes,

          publicKey: provider.wallet.payer.publicKey.toBytes(),
          message: offchainMessage,

          signature: signatureBytes,
        })
      )
      .add(
        await program.methods
          .proposeAndSignAgreement(
            kvPairs,
            provider.wallet.publicKey,
            signature,
          )
          .accounts({
            dataEntry: newDataEntryAccountKp.publicKey,
            signer: provider.wallet.publicKey,
            sysvarIx: web3.SYSVAR_INSTRUCTIONS_PUBKEY,
          })
          .signers([newDataEntryAccountKp])
          .instruction()
      );

    // Send tx
    try {
      // Send and confirm the transaction
      const signature = await web3.sendAndConfirmTransaction(
        provider.connection, // Use provider's connection
        tx, // The transaction
        [
          provider.wallet.payer,
          newDataEntryAccountKp
        ], // Signers
        {
          commitment: "confirmed", // Wait for confirmation
        }
      );

      console.log(`Transaction successful with signature: ${signature}`);

      // Optionally, verify the transaction's effects
      const txDetails = await provider.connection.getTransaction(signature, {
        commitment: "confirmed",
      });
      expect(txDetails).to.not.be.null;
      // Add more assertions based on expected state changes
    } catch (error) {
      console.error("Transaction failed:", error);
      throw error; // Fail the test on error
    }
  });

  it("accepts secp256k1(EVM) signatures", async () => {
    const kvPairs: {key: string, value: string}[] = [
      { key: "name", value: "Alice"},
      { key: "age", value: "30"},
    ];

    // TODO test
    const messageTypes = ["string", "uint16"];
    const messageValues = ["Alice", 123];

    const ethSigner = new ethers.Wallet(process.env.ETH_SIGNER_PRIVATE_KEY);

    // keccak256 hash of the message
    const messageHash = ethers.solidityPackedKeccak256(
      messageTypes,
      messageValues,
    );

    // get hash as Uint8Array of size 32
    const messageHashBytes: Uint8Array = ethers.getBytes(messageHash);

    // Signed message that is actually this:
    // sign(keccak256("\x19Ethereum Signed Message:\n" + len(messageHash) + messageHash)))
    const fullSignature = await ethSigner.signMessage(messageHashBytes);

    let fullSignatureBytes = ethers.getBytes(fullSignature);
    const signatureBytes = fullSignatureBytes.slice(0, 64);
    const recoveryIdBytes = fullSignatureBytes[64] - 27;

    const offchainMessage = Buffer.concat([
      Buffer.from('\x19Ethereum Signed Message:\n32'),
      messageHashBytes,
    ]);

    const newDataEntryAccountKp = new web3.Keypair();
    const tx = (new web3.Transaction())
      // SvmAgreementRegistry will check if the previous instruction is a valid Ed25519 instruction, so we must call it
      // immediately before calling SvmAgreementRegistry.
      .add(
        anchor.web3.Secp256k1Program.createInstructionWithEthAddress({
          message: offchainMessage,
          ethAddress: await ethSigner.getAddress(),
          signature: signatureBytes,
          recoveryId: recoveryIdBytes,
        })
      )
      .add(
        await program.methods
          .proposeAndSignAgreementEth(
            kvPairs,
            Array.from(ethers.getBytes(await ethSigner.getAddress())),
            Array.from(signatureBytes),
            offchainMessage,
            recoveryIdBytes,
          )
          .accounts({
            dataEntry: newDataEntryAccountKp.publicKey,
            signer: provider.wallet.publicKey,
            sysvarIx: web3.SYSVAR_INSTRUCTIONS_PUBKEY,
          })
          .signers([newDataEntryAccountKp])
          .instruction()
      );

    // Send tx
    try {
      // Send and confirm the transaction
      const signature = await web3.sendAndConfirmTransaction(
        provider.connection, // Use provider's connection
        tx, // The transaction
        [
          provider.wallet.payer,
          newDataEntryAccountKp
        ], // Signers
        {
          commitment: "confirmed", // Wait for confirmation
        }
      );

      console.log(`Transaction successful with signature: ${signature}`);

      // Optionally, verify the transaction's effects
      const txDetails = await provider.connection.getTransaction(signature, {
        commitment: "confirmed",
      });
      expect(txDetails).to.not.be.null;
      // Add more assertions based on expected state changes
    } catch (error) {
      console.error("Transaction failed:", error);
      throw error; // Fail the test on error
    }
  });
});

function serializeKeyValuePairs(items: { key: string; value: string }[]): Uint8Array {
  const coder = new BorshCoder({
    "address": "11111111111111111111111111111111",
    "metadata": {
      "name": "placeholder",
      "version": "0.1.0",
      "spec": "0.1.0",
    },
    "instructions": [],
    "accounts": [],
    "types": [
      {
        "name": "KeyValuePairs",
        "type": {
          "kind": "struct",
          "fields": [
            {
              "name": "kvPairs",
              "type": {
                "vec": {
                  "defined": {
                    "name": "KeyValuePair"
                  }
                }
              }
            }
          ]
        }
      },
      {
        "name": "KeyValuePair",
        "type": {
          "kind": "struct",
          "fields": [
            {
              "name": "key",
              "type": "string"
            },
            {
              "name": "value",
              "type": "string"
            }
          ]
        }
      }
    ]
  });

  return coder.types.encode("KeyValuePairs", {
    kvPairs: items,
  });
}

File: tests/src/svm_agreement_registry.rs
use std::{
    env,
    rc::Rc,
};
use alloy::{
    signers::{SignerSync, local::PrivateKeySigner},
    primitives::keccak256,
};
use anchor_lang::solana_program::sysvar;
use anyhow::Result;
use anchor_client::{Client, Cluster, solana_sdk::signature::{read_keypair_file, Keypair, Signer}, Program};
use solana_commitment_config::CommitmentConfig;
use solana_ed25519_program::new_ed25519_instruction_with_signature;
use solana_secp256k1_program::new_secp256k1_instruction_with_signature;
use svm_agreement_registry::{
    DataEntry,
    utils::ed25519::{KeyValuePair, format_message as format_ed25519_message},
    utils::secp256k1::{format_message as format_secp256k1_message},
};

fn setup() -> Result<(Program<Rc<Keypair>>, Rc<Keypair>, Keypair, Vec<KeyValuePair>)> {
    let anchor_wallet = std::env::var("ANCHOR_WALLET")?;
    let payer = Rc::new(read_keypair_file(&anchor_wallet).unwrap());
    let client = Client::new_with_options(Cluster::Localnet, payer.clone(), CommitmentConfig::confirmed());
    let svm_agreement_registry_program = client.program(svm_agreement_registry::ID)?;

    let data_entry = Keypair::new();

    let kv_pairs = vec![
        KeyValuePair {
            key: "name".to_string(),
            value: "Alice".to_string(),
        },
        KeyValuePair {
            key: "age".to_string(),
            value: "30".to_string(),
        },
    ];

    Ok((svm_agreement_registry_program, payer, data_entry, kv_pairs))
}

#[test]
fn test_accepts_ed25519_signatures() -> Result<()> {
    let (svm_agreement_registry_program, payer, data_entry, kv_pairs) = setup()?;

    let offchain_message = format_ed25519_message(&kv_pairs)?;

    let signature = payer.sign_message(&offchain_message);

    let tx = svm_agreement_registry_program
        .request()
        .instruction(
            new_ed25519_instruction_with_signature(
                &offchain_message,
                signature.as_array(),
                &payer.pubkey().to_bytes(),
            ),
        )
        .instruction(
            svm_agreement_registry_program
                .request()
                .accounts(svm_agreement_registry::accounts::StoreData {
                    data_entry: data_entry.pubkey(),
                    signer: svm_agreement_registry_program.payer(),
                    system_program: solana_system_interface::program::ID,
                    sysvar_ix: sysvar::instructions::ID,
                })
                .args(svm_agreement_registry::instruction::ProposeAndSignAgreement {
                    kv_pairs: kv_pairs.clone(),
                    signer: svm_agreement_registry_program.payer(),
                    signature: signature.into(),
                })
                .instructions()?
                .remove(0)
        )
        .signer(&data_entry)
        .send()?;

    println!("Transaction signature: {}", tx);

    let data_entry_account: DataEntry = svm_agreement_registry_program.account(data_entry.pubkey())?;
    assert_eq!(data_entry_account.kv_pairs, kv_pairs);
    assert_eq!(data_entry_account.signer, payer.pubkey());

    Ok(())
}

#[test]
fn test_accepts_secp256k1_signatures() -> Result<()> {
    let (svm_agreement_registry_program, payer, data_entry, kv_pairs) = setup()?;

    let offchain_message = format_secp256k1_message(&kv_pairs)?;

    let eth_signer = env::var("ETH_SIGNER_PRIVATE_KEY")?.parse::<PrivateKeySigner>()?;
    let eth_address = eth_signer.address();
    let full_signature = eth_signer.sign_hash_sync(&keccak256(&offchain_message))?;
    let signature = full_signature.as_erc2098();
    let recovery_id = full_signature.recid().to_byte();

    let tx = svm_agreement_registry_program
        .request()
        .instruction(
            new_secp256k1_instruction_with_signature(
                &offchain_message,
                &signature,
                recovery_id,
                &eth_address.into_array(),
            ),
        )
        .instruction(
            svm_agreement_registry_program
                .request()
                .accounts(svm_agreement_registry::accounts::StoreData {
                    data_entry: data_entry.pubkey(),
                    signer: svm_agreement_registry_program.payer(),
                    system_program: solana_system_interface::program::ID,
                    sysvar_ix: sysvar::instructions::ID,
                })
                .args(svm_agreement_registry::instruction::ProposeAndSignAgreementEth {
                    kv_pairs: kv_pairs.clone(),
                    signer: eth_address.into_array(),
                    signature,
                    recovery_id,
                })
                .instructions()?
                .remove(0)
        )
        .signer(&data_entry)
        .send()?;

    println!("Transaction signature: {}", tx);

    let data_entry_account: DataEntry = svm_agreement_registry_program.account(data_entry.pubkey())?;
    assert_eq!(data_entry_account.kv_pairs, kv_pairs);
    assert_eq!(data_entry_account.signer.to_bytes()[12..32], eth_address.into_array());

    Ok(())
}
